### ISSUE TITLE
Consider splitting yargs into a collection of composable packages

### ISSUE THREAD

**Issue by nexdrew** [yargs/yargs#579]
Based on conversation here: https://github.com/yargs/yargs/pull/562#issuecomment-234155140

I believe we could break up the different features of yargs into separate modules across different packages, satisfying the following goals/suggestions:
1. Keep the standard `yargs` library as an API with all composable features enabled by default (includes each feature-specific package as a dependency), so `yargs` users get practically the same experience they get today
2. Put the bulk of the current `yargs` API in a `yargs-core` (or similar) package with composable/optional features disabled by default
   
   This probably means (a) using an API method for a disabled feature throws an error and (b) features are enabled by giving `yargs-core` an object/module that it can determine fulfills a certain contract (i.e. implements certain methods) representing some portion of the overall API.
3. Split out any or all of the following features into their own `yargs-${feature}` package:
   - i18n strings and locale resolution
   - usage/help content, formatting, and layout
   - error handling
   - command support
   - shell completion support
   - package.json lookups (possibly just an explicit opt-in in `yargs-core`)
4. Allow users to swap `yargs` for `yargs-core` + any feature-specific packages they need to enable a subset of full yargs functionality with a smaller set of dependencies
5. Allow users to plug their own feature-specific package(s) into `yargs-core` to override the default functionality of all or part of a feature

This is no small feat, and it will require decoupling the contracts between current submodules to define a clearer separation of concerns via explicit interfaces, but I think it could be worthwhile to turn yargs into a pluggable platform for building CLIs, similar to how web frameworks like `express` or `restify` support pluggable middleware.

**Comment by maxrimue** [comment]
How's it looking with this feature? What do folks think about it?

Of course it would require a lot of work as pointed out by @nexdrew, but I do think this could be extremely useful. It's hard to satisfy people with speed and functionality simultaneously, this way we could ensure people can build together yargs right how they need it, in an easy way.

Also, it's easier to create new features, and instead of adding more and more to the yargs repo, people could just go ahead and publish their own packages if it's features not everyone/only a few will need. This "You can create your own package" mentality also helps when people all have their own opinions how things should work, or heavily depend on us to maintain extra features.

What I wonder though: Should this `yargs-core` package include the parser or depend on it like yargs does right now? From how it looks like, people use the yargs-parser independently from yargs only when they want to create a lightweight interface. Since that is what `yargs-core` would head for (?), it would make sense to put the parser right in.

So what's everyone's opinion on the general idea of this? Should we head for this in 2017?

**Comment by bcoe** [comment]
@maxrimue @nexdrew I'm of two minds; I actually think we've hit a pretty good sweet spot in terms of performance with yargs:

1. we did some profiling, found a few hot-spots, and I think for the amount of functionality the library is pretty efficient.
2. I think we have found some good division points already for yargs (yargs-parser, y18n, cliui) and it's relatively modular.
3. we cut down on some dependencies, and aren't relying on too many in the grand scheme of things.

tldr; I don't think we should go out of our way to split yargs into too many more components, or to eliminate too many more dependencies in the name of efficiency. _however_, it would be interesting to think of a few more modules to split yargs into; but I think the rule should be, within reason, only sub-divide a component into a module if we think other library developers would find it useful.

**Comment by bcoe** [comment]
we might revisit this as part of the discussion in https://github.com/yargs/yargsa/issues/5, but closing for now since we've hit a relative steady-state (in terms of composition) in yargs/yargs.

**Comment by karfau** [comment]
For me it would be really helpful to be able to just use `yargs-parser` and the `lib/usage` part of yargs. I'm even thinking about making a fork, that only offers that part (could even drop the support for the commands part that I don't need).

I agree that the whole work that was initially suggested sounds like a lot of work, but I think splitting out the usage part would a much simpler task and would be something that people would benefit most from. Even if it still comes with the full commands support, in case that makes things easier.

BTW: the link you provided in the last comment is pointing to github 404 page.
